suite/tests/client-interface/drreg-test.dll.c

```diff
@@ -320,20 +320,30 @@ event_instru2instru(void *drcontext, void *tag, instrlist_t *bb, bool for_trace,
     instr_t *inst = instrlist_first(bb);
     drreg_status_t res;
     drvector_t allowed;
-    reg_id_t reg;
+    reg_id_t reg0, reg1;
 
     drreg_init_and_fill_vector(&allowed, false);
     drreg_set_vector_entry(&allowed, TEST_REG, true);
 
-    res = drreg_reserve_register(drcontext, bb, inst, NULL, &reg);
+    res = drreg_reserve_register(drcontext, bb, inst, NULL, &reg0);
+    CHECK(res == DRREG_SUCCESS, "default reserve should always work");
+    /* We are reserving a register again. This call is made to expose a possible bug
+     * (xref i#3821).
+     */
+    res = drreg_reserve_register(drcontext, bb, inst, NULL, &reg1);
     CHECK(res == DRREG_SUCCESS, "default reserve should always work");
-    res = drreg_unreserve_register(drcontext, bb, inst, reg);
+    instrlist_meta_preinsert(bb, inst,
+                             XINST_CREATE_load_int(drcontext, opnd_create_reg(reg0),
+                                                   OPND_CREATE_INT32(MAGIC_VAL)));
```

Comment:
Still not fully understanding this test: It looks to me like the code is just hoping that there will be consequences of clobbering the reg if it's not restored -- but there's no guarantee.  There could always be either a dead reg or a reg that won't cause a visible crash at that point.  In fact that is probably not unlikely.  A stronger test is needed it seems to me.